### DISCUSSION TITLE
Ellipsis literal in GDScript??

### DISCUSSION THREAD

**Discussion by Meorge** [godotengine/godot-proposals#10191] [Scripting]
Python has the [Ellipsis object](https://docs.python.org/dev/library/constants.html#Ellipsis), which is often used for slicing multi-dimensional containers. Another common use of it, and one that I really like a lot, is as a no-op placeholder in functions you have yet to write code for. Often when I'm blocking out my function structures in a Python program, I'll quickly write out something like
```py
def my_first_function():
    ...
```

Personally, I'd love to be able to do the same thing in GDScript. If I'm adding functions to my classes for handling taking damage, I could type something like
```gdscript
func take_damage(amount: int):
    ...
```
then return to fill in the actual code later. I imagine it'd be equivalent to `pass` in this context, as it is in Python, but:
- It saves one (1) character of typing. Efficiency!
- Since all three characters are the same, you can tap-tap-tap the placeholder out quicker than typing a specific word. Honestly this feels like the biggest benefit of `...` over `pass` to me - when I'm at this stage of coding I want something I can very quickly bang out.
- It's arguably a bit "English-ier" than `pass`, and thus easier for someone not familiar with programming to quickly understand. Ellipsis are used widely in English (and other languages too?) to take the place of words when there's silence. When people see this, they *might* be more likely to intuitively understand that it means there's nothing here/it's a placeholder and stuff should be written here. (Maybe a stretch.)

I think I've spoken to at least one other person who said they would appreciate this bit of syntactic sugar, and maybe there are even more out there? Perhaps if there were other valid uses we could come up with for an Ellipsis literal and/or object, it could make sense to consider further? For now, I mostly just think it'd be really nice to be able to tap-tap-tap in a no-op 😅

**Comment by vvvvvvitor** [comment]
To add to the reasons this should be added: 

it would also allow for variables/objects/signals/etc named "pass".

**Reply by Meorge** [reply]
It could be nice to free up that keyword for use as a variable/etc! However I would be a bit worried about people with existing `pass` statements in their code suddenly being errors for undeclared identifiers. I also think that `pass` is better than `...` when you explicitly want to say "nothing should be done in this block of code". The ellipsis `...` read more to me as "some code should *eventually* go here but I haven't written it yet".

**Comment by dalexeev** [comment]
There is a suggestion to use `...` for variadic functions (rest parameter and spread syntax). See #1034.

As for using `...` as an alternative for `pass`, we prefer not to have multiple syntax options _doing the same thing_. For example, we have two dictionary styles, but the Lua style allows you to omit quotes for keys. We have two setter/getter styles, but the alternative style allows you to use a separate function like `setget` in 3.x.

**Comment by AThousandShips** [comment]
This would be very confusing when writing code examples and pseudocode, how would you denote *actually* omitting part of the code as distinct from `pass`?

I also consider non-word based control statements confusing and unhelpful, `pass` is pretty self evident, or no less so than `func` or `signal` or `@export`, but `...` doesn't really tell me anything about what it does in this context

**Reply by Meorge** [reply]
Good point about `...` being "I actually omitted code here for the purposes of brevity in this snippet, but you should have it" versus "I will write more code here in the future".

I suppose `...` *could* trigger a warning from the VM whenever it was encountered that basically told the user, "Hey, looks like you intended to write something here but never got around to it. You probably want to get that written otherwise the code you've written won't work correctly." But that also feels a bit too extraneous for the core VM, and maybe something that would be better suited for an add-on that scanned comments.

**Comment by miv391** [comment]
If I would see this in any code/text editor
```python
def my_first_function():
    ...
```
my first though would be that this is an editor feature where the function implementation is collapsed, maybe because is it exceptionally long or whatever. Then I would try to find a way to uncollapse the code.
